src/main/java/org/jenkinsci/remoting/engine/JnlpAgentEndpoint.java

```diff
@@ -0,0 +1,258 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.remoting.engine;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.nio.channels.SocketChannel;
+import java.security.interfaces.RSAPublicKey;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.jenkinsci.remoting.util.IOUtils;
+import org.jenkinsci.remoting.util.KeyUtils;
+
+import static org.jenkinsci.remoting.engine.EngineUtil.readLine;
+
+/**
+ * Represents a {@code TcpSlaveAgentListener} endpoint details.
+ *
+ * @since FIXME
+ */
+public class JnlpAgentEndpoint {
+    /**
+     * The hostname to connect to.
+     */
+    @Nonnull
+    private final String host;
+    private final int port;
+    /**
+     * The {@code InstanceIdentity.getPublic()} of the instance or {@code null} if the instance identity was not
+     * published.
+     */
+    @CheckForNull
+    private final RSAPublicKey publicKey;
+    /**
+     * The {@code AgentProtocol.getName()}s supported by the instance or {@code null} if the supported protocols are
+     * not published.
+     */
+    @CheckForNull
+    private final Set<String> protocols;
+
+    /**
+     * Constructor for a remote {@code Jenkins} instance.
+     *
+     * @param host      the hostname.
+     * @param port      the port.
+     * @param publicKey the {@code InstanceIdentity.getPublic()} of the remote instance (if known).
+     * @param protocols The supported protocols.
+     */
+    public JnlpAgentEndpoint(@Nonnull String host, int port, @CheckForNull RSAPublicKey publicKey,
+                             @javax.annotation.CheckForNull Set<String> protocols) {
+        this.host = host;
+        this.port = port;
+        this.publicKey = publicKey;
+        this.protocols = protocols == null ? null : Collections.unmodifiableSet(new LinkedHashSet<String>(protocols));
+    }
+
+    /**
+     * Gets the socket address.
+     *
+     * @return the socket address
+     */
+    @Nonnull
+    public InetSocketAddress getAddress() {
+        return new InetSocketAddress(host, port);
```

Comment:
Port range is not verified in the constructor, we can get IllegalArgumentException here